pythonFiles/vscode_datascience_helpers/dataframes/vscodeDataFrame.py

```diff
@@ -1,11 +1,57 @@
 import pandas as _VSCODE_pd
 import builtins as _VSCODE_builtins
 import json as _VSCODE_json
+import numpy as _VSCODE_np
 import pandas.io.json as _VSCODE_pd_json
 
 # PyTorch and TensorFlow tensors which can be converted to numpy arrays
 _VSCODE_allowedTensorTypes = ["Tensor", "EagerTensor"]
 
+# Convert numpy array to DataFrame
+def _VSCODE_convertNumpyArrayToDataFrame(ndarray):
+    # Save the user's current setting
+    current_options = _VSCODE_np.get_printoptions()
+    # Ask for the full string. Without this numpy truncates to 3 leading and 3 trailing by default
+    _VSCODE_np.set_printoptions(threshold=99999)
+
+    temp = ndarray
+    x_len = temp.shape[0]
+    y_len = None
+    # Figure out if we're dealing with ragged data
+    try:
+        # Handle ragged arrays by making a container where the number of
+        # columns is the max length of all rows. Missing elements are
+        # represented as empty strings.
+        y_len = max([len(temp[i]) for i in range(x_len)])
+    except TypeError:
+        # 1D ndarray, nothing to do here
+        pass
+
+    if y_len:
+        # Figure out what kind of object the rows are
+        flattened = _VSCODE_np.full((x_len, y_len), "", dtype="object")
+        for i in range(x_len):
+            for j in range(len(temp[i])):
+                element = temp[i][j]
+                if isinstance(element, _VSCODE_np.ndarray):
+                    stringified = _VSCODE_np.array2string(element, separator=", ")
```

Comment:
I thought you found this was super slow and you were going to do this in the UI?

Reply:
No, I found that lazy fetching details of individual cells was too slow (2 seconds for long values). Here I'm requesting the full string and then doing all the truncation in the UI.

Reply:
Oh so you're concatenating the value here (instead of sending the full json).

Reply:
Sorry I though the slowdown in the UI meant you were going to do the concat there as well.

Reply:
Bit confused. By 'concatenating' do you mean transforming the data from an array to a string?

Just to be sure we're on the same page, the intent in this bit of code is to generate the full string representation of all data past 2 dimensions, without any truncation, over to the UI. Then in the cellFormatter I'm displaying all values with a trailing ellipsis, so we only render the first 10 characters of any value. We display the full untruncated value later.

In a separate PR I'll also implement pagination like we talked about, which should minimize the performance impact of sending untruncated strings to the UI.

Reply:
Yes I meant transforming the data from an array to a string. Theoretically the json for the array isn't much more than the json for the string so I thought you were just going to send the array itself over. But maybe that is significantly slower.

And I believe you'll be throwing this code away when you implement slicing?

Reply:
> I thought you were just going to send the array itself over

I think this doesn't really work for >3D data. Python objects know how to represent themselves as strings in Python code, but we can't get the string representation out of them in TypeScript. At minimum we'd have to recursively convert Python objects like Tensors or numpy ndarrays to arrays, and then I'm not sure how we'd reconstitute nested arrays.

> you'll be throwing this code away when you implement slicing

Unfortunately not, because the slice data view as currently spec-ed only applies to 3D data. We've talked to Jill about reworking the design, but I believe the current plan is to maybe extend it to work for 2D or 3D data. 

If we made the slice data view work for arbitrary dimensions and selected default dimensions for higher dimensional arrays, then yes we'd be able to deprecate this code. Jeffrey and I have discussed whether the latter would be confusing for users (e.g. they might think the data is 2D because we default them to a flat view with a preselected slice). So the current plan is to show the flattened view (implemented in this PR) by default, and then also provide a way to turn on slicing mode.